src/utils/crypto_utils/src/crypto_lib.c

```diff
@@ -375,3 +378,782 @@ CryptoKeyHandle GetRootKeyForKeyID(const char* kid)
+        outSize = 0;
+    }
+
+    *destBuff = encryptedBuffer;
+
+    return result;
+}
+
+/**
+ * @brief Helper function that loads the @p buffToFill with the offset bytes
+ *
+ * @param buffToFill the buffer to allocated with the bytes of offset
+ * @param blockSize the blockSize for the block who's offset is being loading into the buffer
+ * @param offset the offset to be loaded into @p buffToFill
+ */
+void LoadIvBufferWithOffsetHelper(unsigned char* buffToFill, size_t blockSize, unsigned long long offset)
+{
+    unsigned long long mask = 0xFF;
+    for (int i = blockSize - 1; i >= 0; --i)
+    {
+        buffToFill[i] = mask & offset;
+        mask <<= 8; // shift left 8 bytes to select the next value
+    }
+}
+
+/**
+ * @brief Helper fucntion for calculating the intialization vector for a certain block
+ * @details the IV is calculated as the Encrypt(<offSet-in-a-byte-buffer-of-block-size>, eKey)
+ * @param destIvBuffer the buffer to store the bytes of the iv in
+ * @param blockSize the size of the block for the algorithm, indicates the size that @p destIvBuffer must be
+ * @param alg the algorithm the IV is being generated for
+ * @param eKey the key to use to calculate the intialization vector
+ * @param offSet the offSet for the block who's IV is being calculated
+ * @return a value of ADUC_Result
+ */
+ADUC_Result CryptoUtils_CalculateIV(
+    unsigned char** destIvBuffer, size_t blockSize, const DecryptionAlg alg, const KeyData* eKey, const size_t offSet)
+{
+    ADUC_Result result = { .ResultCode = ADUC_GeneralResult_Failure, .ExtendedResultCode = 0 };
+
+    unsigned char* tempDestIvBuff = NULL;
+    unsigned char* ivByteBuff = NULL;
+
+    if (CryptoUtils_IsKeyNullOrEmpty(eKey) || destIvBuffer == NULL)
+    {
+        goto done;
+    }
+
+    ivByteBuff = (unsigned char*)malloc(blockSize);
+
+    memset(ivByteBuff, 0, blockSize);
+
+    LoadIvBufferWithOffsetHelper(ivByteBuff, blockSize, offSet);
+
+    result = CryptoUtils_EncryptBlock(&tempDestIvBuff, alg, eKey, NULL, ivByteBuff, blockSize);
+
+    if (IsAducResultCodeFailure(result.ResultCode))
+    {
+        goto done;
+    }
+
+    result.ResultCode = ADUC_GeneralResult_Success;
+done:
+
+    if (IsAducResultCodeFailure(result.ResultCode))
+    {
+        free(tempDestIvBuff);
+        tempDestIvBuff = NULL;
+    }
+
+    free(ivByteBuff);
+
+    *destIvBuffer = tempDestIvBuff;
+
+    return result;
+}
+
+/**
+ * @brief Helper function for finding the number of padded bytes in the searchBlock according to PKCS#7
+ *
+ * @param searchBlock the block to search for the padded size
+ * @param blockSize the size of the block
+ * @return the number of bytes that are padding in @p searchBlock
+ */
+size_t FindPKCS7BytesPadded(char* searchBlock, const size_t blockSize)
+{
+    size_t paddingCandidate = searchBlock[blockSize - 1];
+
+    if (paddingCandidate > 0 && paddingCandidate <= 0x10)
+    {
+        size_t index;
+        for (index = blockSize - 1; index >= 0; --index)
```

Comment:
why searching backwards? non-obvious, and some of this math (line 854) makes me nervous.

Reply:
Padding will always exist at the end. For PKCS#7 the minimum number of paddedbytes is 0x01 but we need to check that there is indeed only one 0x01 at the end of the buffer (and there's not 0x10 after that). Same with 0x02, 0x03, etc all the way up to 0x10. The best candidate will always be at the end of the buffer and we count backwards from there to check that we're indeed working with padding and not actual data.